File: dart/dna424_client/test/conformance_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:convert/convert.dart';
import 'package:dna424_client/dna424_client.dart';
import 'package:test/test.dart';

/// Cross-SDK conformance: buildBinding must match the shared vectors (and thus
/// the TS SDK and the Python reference). See sdk/conformance/binding_vectors.json.
void main() {
  test('buildBinding matches the shared conformance vectors', () {
    // test/ -> dna424_client -> dart -> sdk/conformance
    final file = File('${Directory.current.path}/../../conformance/binding_vectors.json');
    final data = jsonDecode(file.readAsStringSync()) as Map<String, dynamic>;
    for (final v in data['vectors'] as List) {
      final got = hex.encode(buildBinding(v['uid'] as String, v['context'] as String));
      expect(got, equals(v['binding_hex']),
          reason: 'binding for ${v['uid']}|${v['context']}');
    }
  });
}

File: ts/veritag-sdk/test/conformance.test.mjs
/**
 * Cross-SDK conformance: buildBinding must match the shared vectors (and thus
 * the Dart SDK and the Python reference). See sdk/conformance/binding_vectors.json.
 */
import test from "node:test";
import assert from "node:assert/strict";
import { readFileSync } from "node:fs";

import { buildBinding } from "../dist/index.js";

const vectors = JSON.parse(
  readFileSync(new URL("../../../conformance/binding_vectors.json", import.meta.url)),
).vectors;

const toHex = (u8) => [...u8].map((b) => b.toString(16).padStart(2, "0")).join("");

test("buildBinding matches the shared conformance vectors", async () => {
  for (const v of vectors) {
    const got = toHex(await buildBinding(v.uid, v.context));
    assert.equal(got, v.binding_hex, `binding for ${v.uid}|${v.context}`);
  }
});
